src/protocols/http/hpr_roaming_protocol.erl

```diff
@@ -0,0 +1,541 @@
+-type prstart_ans() :: map().
+-type xmitdata_req() :: map().
+-type xmitdata_ans() :: map().
+
+-type netid_num() :: non_neg_integer().
+-type gateway_time() :: non_neg_integer().
+
+-type downlink() :: {
+    ResponseStream :: hpr_router_stream_manager:gateway_stream(),
+    Response :: any()
+}.
+
+-type transaction_id() :: integer().
+-type region() :: atom().
+-type token() :: binary().
+-type dest_url() :: binary().
+-type flow_type() :: sync | async.
+
+-define(TOKEN_SEP, <<"::">>).
+
+-record(packet, {
+    packet_up :: hpr_packet_up:packet(),
+    gateway_time :: gateway_time(),
+    response_stream :: hpr_router_stream_manager:gateway_stream()
+}).
+-type packet() :: #packet{}.
+
+-type downlink_packet() :: #packet_router_packet_down_v1_pb{}.
+
+-export_type([
+    netid_num/0,
+    packet/0,
+    gateway_time/0,
+    downlink/0,
+    downlink_packet/0
+]).
+
+%% ------------------------------------------------------------------
+%% Uplink
+%% ------------------------------------------------------------------
+
+-spec new_packet(
+    PacketUp :: hpr_packet_up:packet(),
+    GatewayTime :: gateway_time(),
+    GatewayStream :: hpr_router_stream_manager:gateway_stream()
+) -> #packet{}.
+new_packet(PacketUp, GatewayTime, GatewayStream) ->
+    #packet{
+        packet_up = PacketUp,
+        gateway_time = GatewayTime,
+        response_stream = GatewayStream
+    }.
+
+-spec gateway_time(#packet{}) -> gateway_time().
+gateway_time(#packet{gateway_time = GWTime}) ->
+    GWTime.
+
+-spec response_stream(#packet{}) -> hpr_router_stream_manager:gateway_stream().
+response_stream(#packet{response_stream = ResponseStream}) ->
+    ResponseStream.
+
+-spec make_uplink_payload(
+    NetID :: netid_num(),
+    Uplinks :: list(packet()),
+    TransactionID :: integer(),
+    ProtocolVersion :: pv_1_0 | pv_1_1,
+    DedupWindowSize :: non_neg_integer(),
+    Destination :: binary(),
+    FlowType :: sync | async,
+    RoutingInfo :: hpr_routing:routing_info()
+) -> prstart_req().
+make_uplink_payload(
+    NetID,
+    Uplinks,
+    TransactionID,
+    ProtocolVersion,
+    DedupWindowSize,
+    Destination,
+    FlowType,
+    RoutingInfo
+) ->
+    #packet{
+        packet_up = PacketUp,
+        gateway_time = GatewayTime,
+        response_stream = ResponseStream
+    } = select_best(Uplinks),
+    Payload = hpr_packet_up:payload(PacketUp),
+    PacketTime = hpr_packet_up:timestamp(PacketUp),
+
+    Region = hpr_packet_up:region(PacketUp),
+    DataRate = hpr_packet_up:datarate(PacketUp),
+    Frequency = hpr_packet_up:frequency_mhz(PacketUp),
+
+    {RoutingKey, RoutingValue} =
+        case RoutingInfo of
+            {devaddr, DevAddr} -> {'DevAddr', encode_devaddr(DevAddr)};
+            {eui, DevEUI, _AppEUI} -> {'DevEUI', encode_deveui(DevEUI)}
+        end,
+
+    Token = make_uplink_token(TransactionID, Region, PacketTime, Destination, FlowType),
+
+    ok = hpr_roaming_downlink:insert_handler(TransactionID, ResponseStream),
```

Comment:
I wonder if this should be here. `make_uplink_payload` doesn't imply saving the value in an ets table.

Reply:
This is where it was done in Packet Purchaser. I will look into the reasons for this placement.
cc/ @michaeldjeffrey